### PR TITLE
[SwiftPackageAPIDiff] Adding Initial Tests

### PR DESCRIPTION
**Pull request by goergisn**
## Summary
- Adding tests

### CODE REVIEW

**Comment by atmamont** [comment]
Looks promising! 🤣 
image

**Comment by goergisn** [comment]
> Looks promising! 🤣 
> 
> image
> 
> 

Yeah that's the testing resources mostly 🙈

File: .github/workflows/detect_api_changes.yml [hunk 1]
```diff
@@ -3,11 +3,19 @@ name: Detect public API changes
...
+        required: true
+      old:
+        description: 'The branch & repository of the comparison version "${targetBranch}~${repository}"'
+        required: true
       
 jobs:
 
   build:
-    runs-on: macos-14-xlarge # Apple Silicon Runner
+    runs-on: macos-14 # Apple Silicon Runner
```
Comment (atmamont) [inline .github/workflows/detect_api_changes.yml:18]: Does this change aim for costs optimization?

File: Scripts/public-api-diff/Sources/Helpers/SDKAnalyzer/SDKAnalyzer.swift [hunk 2]
```diff
@@ -58,139 +64,69 @@ enum SDKAnalyzer {
...
-        try allTargets.forEach { targetName in
-            let diff = try SDKAnalyzer.diffSdkDump(
-                newDump: newDumpGenerator.generate(for: targetName),
-                oldDump: oldDumpGenerator.generate(for: targetName)
+        allTargets.forEach { targetName in
+            
+            var newDump: SDKDump? = nil
+            var oldDump: SDKDump? = nil
+            
+            print("🕵️‍♂️ Analyzing `\(targetName)`")
```
Comment (atmamont) [inline Scripts/public-api-diff/Sources/Helpers/SDKAnalyzer/SDKAnalyzer.swift:108]: `OSLog` could be a [nice improvement](https://www.avanderlee.com/debugging/oslog-unified-logging/#:~:text=OSLog%20is%20a%20replacement%20for,Xcode%2015's%20new%20logging%20console.)
Overall, I'd recommend wrapping every output to a `log` method so you can easily switch how exactly you do this in the future by changing only a couple of lines

File: Scripts/public-api-diff/Tests/PublicApiDiffTests/FileHandlingTests.swift [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+        do {
+            try fileHandler.write(output, to: filePath)
+            XCTFail("Write should have thrown an error")
+        } catch {
+            let fileHandlerError = try XCTUnwrap(error as? FileHandlerError)
+            XCTAssertEqual(fileHandlerError, FileHandlerError.couldNotCreateFile(outputFilePath: filePath))
+        }
+    }
+    
+    func test_createCleanRepository() throws {
```
Comment (atmamont) [inline Scripts/public-api-diff/Tests/PublicApiDiffTests/FileHandlingTests.swift]: I would also add a failing test, for instance when the directory couldn't be erased/created due to lack of access

File: Scripts/public-api-diff/Tests/PublicApiDiffTests/GitTests.swift [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+//
+// This file is open source and available under the MIT license. See the LICENSE file for more info.
+//
+
+@testable import public_api_diff
+import XCTest
+
+class GitTests: XCTestCase {
+    
+    func test_clone_success() throws {
```
Comment (atmamont) [inline Scripts/public-api-diff/Tests/PublicApiDiffTests/GitTests.swift:12]: Talking of Git I wonder if we can have any blockers like having some CL prompts to trust the host or to provide additional authorization. Then the script will be timed out I guess?

File: Scripts/public-api-diff/Tests/PublicApiDiffTests/OutputGeneratorTests.swift [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+
+        ---
+        ## `Target_1`
+        - ❇️  Some Addition
+
+        ---
+        **Analyzed modules:** Target_1
+        """
+        
+        let outputGenerator = OutputGenerator(
```
Comment (atmamont) [inline Scripts/public-api-diff/Tests/PublicApiDiffTests/OutputGeneratorTests.swift:47]: this is a nice testable decoupling of business and presentation logic, pleasure to review

File: Scripts/public-api-diff/Tests/PublicApiDiffTests/OutputGeneratorTests.swift [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+        - 🔀 Another Change
+        - 😶‍🌫️ Another Removal
+        ### `Parent_In_Target_2`
+        - 🔀 Another Change
+
+        ---
+        **Analyzed modules:** Target_1, Target_2
+        """
+
+        let outputGenerator = OutputGenerator(
```
Comment (atmamont) [inline Scripts/public-api-diff/Tests/PublicApiDiffTests/OutputGeneratorTests.swift:80]: What do you think of `TextOutputGenerator` or `ConsoleOutputGenerator`? Maybe later on you want to add something like `HTMLOutputGenerator`

**Review by atmamont — DISMISSED** [review]
This is shaping out nicely!

File: .github/workflows/detect_api_changes.yml (hunk 1 above)
Comment (goergisn) [reply .github/workflows/detect_api_changes.yml:18]: Initially it was to get the script running during our small downtime of paid runners - but as it's fast enough on the "smaller" runner we can as well just keep it here :)

File: Scripts/public-api-diff/Sources/Helpers/SDKAnalyzer/SDKAnalyzer.swift (hunk 2 above)
Comment (goergisn) [reply Scripts/public-api-diff/Sources/Helpers/SDKAnalyzer/SDKAnalyzer.swift:108]: Yeah agree - would also allow us to silence the output or have a more verbose output - will add it to the list of improvements

File: Scripts/public-api-diff/Tests/PublicApiDiffTests/OutputGeneratorTests.swift (hunk 6 above)
Comment (goergisn) [reply Scripts/public-api-diff/Tests/PublicApiDiffTests/OutputGeneratorTests.swift:80]: `MarkdownOutputGenerator` would be a good name for the current one - added it to the list of improvements

File: Scripts/public-api-diff/Tests/PublicApiDiffTests/FileHandlingTests.swift (hunk 3 above)
Comment (goergisn) [reply Scripts/public-api-diff/Tests/PublicApiDiffTests/FileHandlingTests.swift]: Good one, added 👍

File: Scripts/public-api-diff/Tests/PublicApiDiffTests/GitTests.swift (hunk 4 above)
Comment (goergisn) [reply Scripts/public-api-diff/Tests/PublicApiDiffTests/GitTests.swift:12]: I did not run into any of this so far - probably there is more to be done for private repositories - might cross the bridge once we get there :P

**Review by atmamont — APPROVED** [review]
🎉

**Review by erenbesel — APPROVED** [review]

**Pull request merged**